### ISSUE TITLE
lnd: enforce using IP/localhost on peer-/rpc- and restport options

### ISSUE THREAD

**Issue by ndeet** [lightningnetwork/lnd#734]
When you forget or omit IP or localhost in those options:
- `--listen`
- `--rpclisten`
- `--restlisten`

You get really strange IP bindings, e.g. forgot `localhost` on --restlisten + --rpclisten:
```
lnd --listen=localhost:10111 --rpclisten=10001 --restlisten=8001 --bitcoin.active --bitcoin.simnet 
listen tcp 0.0.39.17:10009: bind: can't assign requested address
```
It wants to bind to IP `0.0.39.17` and also some random port `10009` (which was never given above)

**Option 1:**
Enforce/abort if there is no "localhost" or IP address given in the *listen options.

**Option 2:**
Default to `localhost` (but the wrong port needs to be fixed) so that this is possible without weird behaviour:
```
lnd --listen=10111 --rpclisten=10001 --restlisten=8001 --bitcoin.active --bitcoin.simnet 
```

**Comment by Roasbeef** [comment]
If you omit the listening interface, you're using the arguments incorrectly. We require the listening interface to be completely defined.

**Comment by ndeet** [comment]
Ok, if it is incorrect, why not throw an error instead of defaulting to some wird 0.0.x.x IP address and cause confusion?

See Slack, some others ran into the same issue already. The previous parameters (`peerport, rpcport, restport`) where used without interface. Seems some like me just replaced them with the new parameters without adding interface to all but `--listen` parameter.

**Comment by minecraft2048** [comment]
Agreed with @ndeet , this is why I used to use the (admittedly) dangerous ` --rpclisten=0.0.0.0:10002 --listen=0.0.0.0:10012 --restlisten=0.0.0.0:8002 --no-macaroons` in https://github.com/lightningnetwork/lnd/issues/672 because before the update we can just define the listening ports without IP/interface using `--rpclisten=10002 --listen=10012` following http://dev.lightning.community/tutorial/01-lncli/index.html . Good thing that tutorial page has been updated to follow the new correct assignment command

**Comment by Roasbeef** [comment]
> The previous parameters (peerport, rpcport, restport) where used without interface.

They were used without an interface, as they all ended in _port_  ;)

>  why not throw an error instead of defaulting to some wired 0.0.x.x IP address and cause confusion

It _does_ throw an error atm, as when you go to bind, you won't be able to actually start to listen on that interface.

**Comment by Roasbeef** [comment]
Defaulting to `localhost` is a sane fallback though.

**Comment by ndeet** [comment]
> They were used without an interface, as they all ended in port ;)
Absolutely correct, nevertheless it was port only before and some people like me just replace the options and oversee that there is also the interface mandatory now :) (agree that this is only temporary problem for users before that change happened)

> It does throw an error atm, as when you go to bind, you won't be able to actually start to listen on that interface.
I agree with you it already throws error. It is just a bit confusing imo. So it would be better UX if it would not fail with something confusing like this:
`listen tcp 0.0.39.17:10009: bind: can't assign requested address`
- maybe the error should be thrown before some strange public IP + port is assigned
- or as you seem to agree fallback to localhost + issuing a warning maybe that

**Comment by Roasbeef** [comment]
Fixed by #877.

**Comment by awb99** [comment]
I have the following issue:  I setup a docker-compose network. And I can only connect to LND via localhost.

This is how I start LND:

```
exec lnd \
    --debuglevel=trace \
    --noseedbackup \
    --logdir="/data" \
    "--bitcoin.active" \
    "--bitcoin.mainnet" \
    "--bitcoin.node"="btcd" \
    "--btcd.rpccert"="/rpc/rpc.cert" \
    "--btcd.rpchost"="blockchain" \
    "--btcd.rpcuser"="$RPCUSER" \
    "--btcd.rpcpass"="$RPCPASS" \
    "--rpclisten=0.0.0.0:10001" \
    "--restlisten=0.0.0.0:8089" \
    "--listen=0.0.0.0:10011" \
```

This is the docker compose:
``` 
version: '2'

services:

    # btc is an image of bitcoin node which used as base image for btcd and
    # btccli. The environment variables default values determined on stage of
    # container start within starting script.
    btc:
      image: btcd
      build:
        context: btcd/
      volumes:
            - shared:/rpc
            - bitcoin:/data
      environment:
        - RPCUSER
        - RPCPASS
        - NETWORK

    btcd:
        extends: btc
        container_name: btcd
        environment:
          - DEBUG
          - MINING_ADDRESS
        entrypoint: ["./start-btcd.sh"]

    btcctl:
        extends: btc
        container_name: btcctl
        links:
            - "btcd:rpcserver"
        entrypoint: ["./start-btcctl.sh"]

    #
    # lnd - lightening daemon
    # 

    lnd:
        image: lnd
        container_name: lnd
        build:
          context: ../
          dockerfile: docker/lnd/Dockerfile
        environment:
          - RPCUSER
          - RPCPASS
          - NETWORK
          - CHAIN
          - DEBUG
        volumes:
            - shared:/rpc
            - lnd:/root/.lnd
        links:
          - "btcd:blockchain"
        ports:
          - "8089:8089"
          - "10001:10001"
          - "10011:10011"
        # expose is not visible from the host
        # expose:
        #  - "8089"
        #  - "10001"
        entrypoint: ["./start-lnd.sh"]
  

    nodeapp:
        image: nodeapp
        container_name: nodeapp
        build:
          context: nodeapp/
        environment:
          - RPCUSER
          - RPCPASS
          - NETWORK
          - CHAIN
          - DEBUG
        volumes:
            - shared:/rpc
            - lnd:/root/.lnd
        # HOST:CONTAINER
        ports:
           - "8088:8088"
        links:
          - "lnd:lnd"
        # expose only works in containers networks; not visible from host
        # expose:
        #  - "8088"
```

I am doing the 0.0.0.0:PORT because I am unsure what IPs docker will give me on app start.
Should LND accept connections via 0.0.0.0 on all network interfaces?? Or should It NOT do so?
